### ISSUE TITLE
Allow to set configuration based on conditions 

### ISSUE THREAD

**Issue by GustavoCaso** [dry-rb/dry-configurable#54]
Example:

```ruby
class App
  extend Dry::Configurable

  setting :logger
  setting :production, if: -> { ENV['RACK_ENV'] == 'production' } do
	end
end

App.configure do |config|
  config.logger = { level: :debug }

  config.production do
    config.logger = { level: :info }
  end
end
```

**Comment by timriley** [comment]
@GustavoCaso I do feel this would be a helpful feature.

However, I wonder if a way for this to be more flexible is to make it work on a pair of condition names and values?

e.g.

```ruby
App.configure do |config|
  config.environment(:production) do
    config.log_level = :info
  end
end
```

Even with the case of environments/"modes" like this, it's clear that a fixed set of names would be limiting, since sometimes people e.g. want to customise their environment names. Would it be better if we declared in this example that `environment` is the condition's _name_, and it has a _value_ that then determines whether the nested settings apply?

I also don't quite understand the top part of your code example... we might want to flesh out a better way of (a) defining which "conditions" are available, along with a default _value_ for that condition, but also (b) being explicit around finalising the config for and passing a particular value for each condition, e.g. something like `finalize(environment: "my_custom_one")`

WDYT?

**Comment by GustavoCaso** [comment]
Hi @timriley 

Thanks for your comment. Yes, I agree with you it would be much better to have each pair of condition names and values.

Sorry, I issue description wasn't very descriptive, I just dumped what had on my head without much thought about it.

Following with your thoughts, what about allowing the user to define there own conditions as a block, that later will be evaluated with the value they pass to it. In your example is `:production` but it could be whatever the user decide.

Something like:

```ruby
class App
  extend Dry::Configurable

  condition :environment do |value = :development|
    ENV['DRY_ENV'] == value
  end
end
```

Later when finalizing the config it will be evaluated:

```ruby
App.configure do |config|
  config.environment(:development) do
    config.log_level = :debug
  end

  config.environment(:production) do
    config.log_level = :info
  end
end
```

I don't how this fits into your thoughts and especially in your `b` point
> being explicit around finalizing the config for and passing a particular value for each condition, e.g. something like finalizing (environment: "my_custom_one")

In the meantime, I started working in POC, but looking forward to hearing your thoughts.
